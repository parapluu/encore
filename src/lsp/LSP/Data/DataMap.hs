module LSP.Data.DataMap(
  LSPData,
  DataMap,
  hasErrorInDataMap
  )
where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Standard
import Data.Map as Map hiding (foldr)

-- LSP
import LSP.Data.Program
import LSP.Data.TextDocument

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

type LSPData = (Program, TextDocument)

type DataMap = Map String LSPData

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

hasErrorInDataMap :: DataMap -> Bool
hasErrorInDataMap dataMap = Map.foldl (\acc (prog, _) -> acc || (length (errors prog)) > 0) False dataMap
