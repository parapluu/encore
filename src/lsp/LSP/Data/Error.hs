module LSP.Data.Error (
  Error(..),
  ErrorType(..),
  fromTCError,
  fromTCErrors,
  fromTCWarning,
  fromTCWarnings,
  fromParsecError,
  fromErrorMessage
) where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Library
import Text.Megaparsec

-- Standard
import qualified Data.Map.Strict as Map
import qualified Data.List.NonEmpty as NE(head)
import Data.List
import Control.Monad

-- Encore imports
import Typechecker.TypeError hiding (Error)
import AST.AST
import qualified AST.Meta as ASTMeta(Position(SingletonPos, RangePos))

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

{- Position tuple  -}
type Position = ((Int, Int), (Int, Int))

data ErrorType = TError | TWarning | THint

instance Show ErrorType where
  show (TError) = "Error"
  show (TWarning) = "Warning"
  show (THint) = "Hint"

{- Error datastructure. Can represent a variety of different error types -}
data Error = Error{
    message     :: String,      -- Error message string
    position    :: Position,    -- Start and end position
    errorType   :: ErrorType
} deriving (Show)

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

extractTCErrorPosition :: TCError -> Position
extractTCErrorPosition error@(TCError errorType []) = ((0, 0), (0, 0))
extractTCErrorPosition error@(TCError errorType backtrace) =
    case fst (head backtrace) of
        (ASTMeta.SingletonPos startPos) ->
            ((getPosLine startPos - 1, getPosCol startPos - 1),
            (getPosLine startPos - 1, getPosCol startPos - 1))
        (ASTMeta.RangePos startPos endPos) ->
            ((getPosLine startPos - 1, getPosCol startPos - 1),
            (getPosLine endPos - 1, getPosCol endPos - 1))
    where getPosLine = fromIntegral . unPos . sourceLine
          getPosCol  = fromIntegral . unPos . sourceColumn

extractTCWarningPosition :: TCWarning -> Position
extractTCWarningPosition warning = ((0,0), (0,0))

fromTCError :: TCError -> Error
fromTCError error =
    Error{message = show error, position = extractTCErrorPosition error, errorType = TError}

fromTCErrors :: [TCError] -> [Error]
fromTCErrors errors = fmap (fromTCError) errors

fromTCWarning :: TCWarning -> Error
fromTCWarning warning =
    Error{message = show warning, position = extractTCWarningPosition warning, errorType = TWarning}

fromTCWarnings :: [TCWarning] -> [Error]
fromTCWarnings warnings = fmap (fromTCWarning) warnings

fromParsecError :: (ParseError Char Text.Megaparsec.Dec) -> Error
fromParsecError error =
    Error{
        message = show error,
        position = ((fromIntegral(unPos $ sourceLine $ NE.head $ errorPos error), fromIntegral(unPos $ sourceColumn $ NE.head $ errorPos error)),
                    (fromIntegral(unPos $ sourceLine $ NE.head $ errorPos error), fromIntegral(unPos $ sourceColumn $ NE.head $ errorPos error))),
        errorType = TError
    }

{- Construct a Error from a message, position and whether it is actually a
    warning

    Param: Error message
    Param: Position (row, column)
    Param: Whether or not the error is actually a warning
-}
fromErrorMessage :: String -> Position -> ErrorType -> Error
fromErrorMessage _message _position _errorType =
    Error{message = _message, position = _position, errorType = _errorType}
