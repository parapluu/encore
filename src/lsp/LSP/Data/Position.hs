module LSP.Data.Position (
  Position,
  Range,
  inRange,
  fromSourcePos,
  fromSourcePosRange,
  rangeCompliment,
  widestRange
) where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Library
import Text.Megaparsec(SourcePos,unPos,sourceLine,sourceColumn)
import Data.List as List

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

{-  -}
type Position = (Int, Int)

{-  -}
type Range = (Position, Position)

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

{-  -}
inRange :: Position -> Range -> Bool
inRange pos (start, end)
  | fst pos > fst start && fst pos < fst end = True
  | fst pos == fst start && fst pos < fst end && snd pos >= snd start = True
  | fst pos > fst start && fst pos == fst end && snd pos < snd end = True
  | fst pos == fst start && fst pos == fst end && snd pos >= snd start && snd pos < snd end = True
  | True = False -- sorry, had to do it (otherwise == True)

{-  -}
fromSourcePos :: SourcePos -> Position
fromSourcePos pos = (fromIntegral (unPos (sourceLine pos)), fromIntegral (unPos (sourceColumn pos)))

{-  -}
fromSourcePosRange :: SourcePos -> SourcePos -> Range
fromSourcePosRange from to = (fromSourcePos from, fromSourcePos to)

posLess :: Position -> Position -> Bool
posLess a b
  | fst a < fst b = True
  | fst a == fst b && snd a < snd b = True
  | otherwise = False

posGreater :: Position -> Position -> Bool
posGreater a b
  | fst a > fst b = True
  | fst a == fst b && snd a > snd b = True
  | otherwise = False

rangeCompliment :: Range -> Range -> Range
rangeCompliment a b
  | posLess (fst a) (fst b) = (fst a, fst b)
  | posGreater (snd a) (snd b) = (snd b, snd a)
  | otherwise = a

minPosition :: [Position] -> Position
minPosition [] = (0, 0)
minPosition (x:[]) = x
minPosition (x:y:xs)
    | posLess x y = minPosition (x:xs)
    | otherwise = minPosition (y:xs)

maxPosition :: [Position] -> Position
maxPosition [] = (0, 0)
maxPosition (x:[]) = x
maxPosition (x:y:xs)
    | posGreater x y = maxPosition (x:xs)
    | otherwise = maxPosition (y:xs)

widestRange :: Range -> [Range] -> Range
widestRange firstRange ranges = do
  let wholeRange = (firstRange:ranges)
  (minPosition (fmap fst wholeRange), maxPosition (fmap snd wholeRange))
