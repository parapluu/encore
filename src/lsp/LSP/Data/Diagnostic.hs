{-# LANGUAGE OverloadedStrings #-}

module LSP.Data.Diagnostic (
    PublishDiagnosticsParams(..),
    Diagnostic(..),
    DiagnosticSeverity(..),
    errorToDiagnostic
) where

import Data.Aeson

import LSP.Data.Position
import LSP.Data.Error

data PublishDiagnosticsParams = PublishDiagnosticsParams {
    pdpUri :: String,
    pdpDiagnostics :: [Diagnostic]
} deriving (Show)

data Diagnostic = Diagnostic {
    dRange :: Range,
    dSeverity :: DiagnosticSeverity,
    dCode :: String,
    dSource :: String,
    dMessage :: String
} deriving (Show)

data DiagnosticSeverity
    = DError
    | DWarning
    | DInformation
    | DHint
    deriving (Show, Enum)

errorToDiagnostic :: Error -> Diagnostic
errorToDiagnostic error =
    Diagnostic {
        dRange    = position error,
        dCode     = "-- error --",
        dSource   = "Encore",
        dMessage  = message error,
        dSeverity = case errorType error of
            TError   -> DError
            TWarning -> DWarning
            THint    -> DHint
    }

{-instance Enum DiagnosticSeverity where
    toEnum = fromIntegral . (-1)
    fromEnum = (+1) . fromInteger-}

instance ToJSON PublishDiagnosticsParams where
    toJSON pdp =
        object [
                "uri"         .= pdpUri pdp,
                "diagnostics" .= pdpDiagnostics pdp
            ]

instance ToJSON Diagnostic where
    toJSON diagnostic =
        object [
                "range"    .= object [
                        "start" .= object [
                                "line"      .= (fst $ fst $ dRange diagnostic),
                                "character" .= (snd $ fst $ dRange diagnostic)
                            ],
                        "end"   .= object [
                                "line"      .= (fst $ snd $ dRange diagnostic),
                                "character" .= (snd $ snd $ dRange diagnostic)
                            ]
                    ],
                "severity" .= (1 + (fromEnum $ dSeverity diagnostic)),
                "code"     .= dCode diagnostic,
                "source"   .= dSource diagnostic,
                "message"  .= dMessage diagnostic
            ]