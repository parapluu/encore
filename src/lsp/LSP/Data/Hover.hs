{-# LANGUAGE OverloadedStrings #-}

module LSP.Data.Hover where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --
import LSP.Data.Position
import Data.Aeson

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --
data TextDocumentPositionParams = TextDocumentPositionParams {
    uri :: String,
    position :: Position
}

data Hover = Hover {
    contents :: String,
    range :: Range
}

-- ###################################################################### --
-- Section: Type Classes
-- ###################################################################### --

instance FromJSON TextDocumentPositionParams where
    parseJSON = withObject "params" $ \o -> do
        textDocument    <- o .: "textDocument"
        uri             <- textDocument .: "uri"
        position        <- o .: "position"
        line            <- position .: "line"
        character       <- position .: "character"

        return TextDocumentPositionParams {
            uri = uri,
            position = (line + 1, character + 1)
        }

instance ToJSON Hover where
    toJSON hover =
        object [
            "contents" .= contents hover,
            "range"    .= object [
                "start" .= object [
                    "line"      .= ((fst $ fst $ range hover) - 1),
                    "character" .= ((snd $ fst $ range hover) - 1)
                ],
                "end"   .= object [
                    "line"      .= ((fst $ snd $ range hover) - 1),
                    "character" .= ((snd $ snd $ range hover) - 1)
                ]
            ]
        ]
