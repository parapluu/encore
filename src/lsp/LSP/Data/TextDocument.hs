{-# LANGUAGE OverloadedStrings #-}

module LSP.Data.TextDocument (
    TextDocument(..),
    TextDocumentClose(..),
    TextDocumentChange(..),
    TextDocumentIdent(..),
    uri,
    version,
    applyTextDocumentChange,
    makeBlankTextDocument
) where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Standard
import Data.Aeson
import Data.String.Utils (split, join)

-- LSP
import LSP.Data.Position

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

data TextDocument = TextDocument {
    tdUri :: String,
    tdVersion :: Int,
    tdLanguageId :: String,
    tdContents :: String
} deriving (Show)

data TextDocumentClose = TextDocumentClose {
    tdclIdentifier :: TextDocumentIdent
} deriving (Show)

data TextDocumentChange = TextDocumentChange {
    tdcIdentifier :: TextDocumentIdentVersion,
    tdcChanges :: [TextDocumentContentChange]
} deriving (Show)

data TextDocumentIdent = TextDocumentIdent {
    tdiUri :: String
} deriving (Show)

data TextDocumentIdentVersion = TextDocumentIdentVersion {
    tdivUri :: String,
    tdivVersion :: Int
} deriving (Show)

data TextDocumentContentChange = TextDocumentContentChange {
    tdccRange :: Range,
    tdccRangeLength :: Int,
    tdccText :: String
} deriving (Show)

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

applyTextDocumentChange :: TextDocumentChange -> TextDocument -> TextDocument
applyTextDocumentChange textDocumentChange textDocument =
    foldr (\change textDocument ->
            TextDocument {
                tdUri        = tdUri textDocument,
                tdVersion    = tdivVersion $ tdcIdentifier textDocumentChange,
                tdLanguageId = tdLanguageId textDocument,
                tdContents   = applyTextChange (tdccRange change)
                                               (tdccText change)
                                               (tdContents textDocument)
            }
        ) textDocument $ tdcChanges textDocumentChange

applyTextChange :: Range -> String -> String -> String
applyTextChange _ replacement "" = replacement
applyTextChange a@((startLine, startChar), (endLine, endChar))
                replacement
                text
    = let textLines = split "\n" text
          clamp = max 0 . min (length textLines - 1)
          boundedStartLine = clamp startLine
          boundedEndLine   = clamp endLine
          startSegment = join "\n" $ take startLine textLines ++
                                     [take startChar $ textLines !! boundedStartLine]
          endSegment = join "\n" $ (drop endChar $ textLines !! boundedEndLine) :
                                   drop (endLine+1) textLines
      in startSegment ++ replacement ++ endSegment

makeBlankTextDocument :: String -> TextDocument
makeBlankTextDocument name = TextDocument{
    tdUri      = name,
    tdVersion  = 1,
    tdLanguageId = "encore",
    tdContents   = ""
}

-- ###################################################################### --
-- Section: Type Classes
-- ###################################################################### --

class TextDocumentURI a where
    uri :: a -> String
class TextDocumentVersion a where
    version :: a -> Int

instance TextDocumentURI TextDocument where
    uri = tdUri
instance TextDocumentVersion TextDocument where
    version = tdVersion

instance TextDocumentURI TextDocumentChange where
    uri = uri . tdcIdentifier
instance TextDocumentVersion TextDocumentChange where
    version = version . tdcIdentifier

instance TextDocumentURI TextDocumentIdent where
    uri = tdiUri

instance TextDocumentURI TextDocumentIdentVersion where
    uri = tdivUri
instance TextDocumentVersion TextDocumentIdentVersion where
    version = tdivVersion

instance FromJSON TextDocument where
    parseJSON = withObject "params" $ \o -> do
        document <- o .: "textDocument"
        uri      <- document .: "uri"
        version  <- document .: "version"
        id       <- document .: "languageId"
        text     <- document .: "text"
        return TextDocument {
            tdUri = uri,
            tdVersion = version,
            tdLanguageId = id,
            tdContents = text
        }

instance FromJSON TextDocumentClose where
    parseJSON = withObject "params" $ \o -> do
        identifier <- o .: "textDocument"

        return TextDocumentClose {
            tdclIdentifier = identifier
        }

instance FromJSON TextDocumentChange where
    parseJSON = withObject "params" $ \o -> do
        identifier  <- o .: "textDocument"
        changes     <- o .: "contentChanges"

        return TextDocumentChange {
            tdcIdentifier = identifier,
            tdcChanges = changes
        }

instance FromJSON TextDocumentContentChange where
    parseJSON = withObject "TextDocumentChange" $ \o -> do
        range       <- o .: "range"
        rangeStart  <- range .: "start"
        startLine   <- rangeStart .: "line"
        startChar   <- rangeStart .: "character"
        rangeEnd    <- range .: "end"
        endLine     <- rangeEnd .: "line"
        endChar     <- rangeEnd .: "character"
        rangeLength <- o .: "rangeLength"
        text        <- o .: "text"

        return TextDocumentContentChange {
            tdccRange = ((startLine, startChar), (endLine, endChar)),
            tdccRangeLength = rangeLength,
            tdccText = text
        }

instance FromJSON TextDocumentIdent where
    parseJSON = withObject "textDocumentIdentifier" $ \o -> do
        uri <- o .: "uri"
        return TextDocumentIdent {
            tdiUri = uri
        }

instance FromJSON TextDocumentIdentVersion where
    parseJSON = withObject "textDocumentIdentifier" $ \o -> do
        uri <- o .: "uri"
        version <- o .: "version"
        return TextDocumentIdentVersion {
            tdivUri = uri,
            tdivVersion = version
        }
