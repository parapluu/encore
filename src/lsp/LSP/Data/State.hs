module LSP.Data.State (
    LSPState(..),
    initial,
    lookupTextDocument,
    addTextDocument,
    closeTextDocument,
    changeTextDocument,
    produceTextDocument,
    compileDocument,
    getProgram
) where

-- ###################################################################### --
-- Section: Imports
-- ###################################################################### --

-- Standard
import Data.Map as Map hiding (foldr)

-- LSP
import LSP.Data.TextDocument
import LSP.Data.Program
import LSP.Data.DataMap
import LSP.Producer

-- ###################################################################### --
-- Section: Data
-- ###################################################################### --

data LSPState = LSPState {
    programs :: DataMap
} deriving (Show)

-- ###################################################################### --
-- Section: Functions
-- ###################################################################### --

initial :: LSPState
initial = LSPState {
    programs = Map.empty
}

lookupTextDocument :: String -> LSPState -> Maybe (TextDocument)
lookupTextDocument uri state =
  case Map.lookup uri (programs state) of
    Just res -> Just $ snd res
    Nothing -> Nothing

addTextDocument :: TextDocument -> LSPState -> LSPState
addTextDocument newDocument (LSPState programs) =
    LSPState (Map.insert (uri newDocument) (makeBlankProgram (uri newDocument), newDocument) programs)

closeTextDocument :: TextDocumentIdent -> LSPState -> LSPState
closeTextDocument ident = LSPState . Map.delete (uri ident) . programs

changeTextDocument :: TextDocumentChange -> LSPState -> LSPState
changeTextDocument documentChange state@(LSPState programs) =
    case Map.lookup (uri documentChange) programs of
        Nothing -> state
        Just (program, textDocument) ->
            LSPState $ Map.insert (uri documentChange)
                                  (program, applyTextDocumentChange documentChange textDocument)
                                  programs

produceTextDocument :: TextDocument -> LSPState -> IO (LSPState)
produceTextDocument textDocument state = do
  let path = uri textDocument
  dataMap <- produceAndUpdateState path (programs state)
  return (LSPState {programs = dataMap})

compileDocument :: String -> LSPState -> IO LSPState
compileDocument path = fmap LSPState . produceAndUpdateState path . programs

getProgram :: String -> LSPState -> Maybe Program
getProgram path = fmap fst . Map.lookup path . programs
