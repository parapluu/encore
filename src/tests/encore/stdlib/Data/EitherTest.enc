
import Data.Either


active class Main
  def main() : unit
    val l = Left[String, int]("error")
    val r = Right[String, int](12)
    l.map(fun (x : int)
            println("should not print")
            x
          end)
    l.foreach(fun (x : int) => println("should not print"))
    r.map(fun (x : int)
            println("should print")
            x
          end)
    r.foreach(fun (x : int) => println("should print"))
    
    match r with 
      case Right(x) => println("should print")
    end
    
    -- TODO: test orElse
    -- TODO: test other functions
  end
end
