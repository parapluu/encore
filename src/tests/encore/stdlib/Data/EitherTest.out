should print
should print
should print
